Added Custom Checker: unmentioned-parameter

CulmoneY opened the pull request:
## Proposed Changes

This checker verifies that each function parameter is explicitly mentioned by name in the function's docstring. If any parameter is missing, the checker raises a warning.

Key Features:
 - Checks all function parameters: Ensures that each parameter passed to a function is documented in the corresponding docstring.
- Handles optional parameters: Parameters with default values are also checked to ensure they are mentioned.
- Handles functions without parameters: The checker ensures no unnecessary warnings are raised for functions with no parameters.
- Manages empty and missing docstrings: Raises warnings when the function contains parameters, but the docstring is empty or missing.
 - If a function's parameter name, followed by a period or a comma, appears in the docstring, the checker will  raise **no** error.
```
def func(x: int):
        """I have named the parameter x."""
        pass
```
- If a function parameter name appears exclusivly within the doctest, the checker will raise an error.

```
def func(x: int) -> list:
         """I do not mention the parameter by name
         
              >>> f(10)
              [x for x in range(10)]
         """
         pass
```

Error messages will look like the following:
<img width="1270" alt="Screenshot 2024-10-01 at 11 31 25 PM" src="https://github.com/user-attachments/assets/0c3989bd-2f63-4af5-803a-e5ffe2502f26">
         
   
## Type of Change

_(Write an `X` or a brief description next to the type or types that best describe your changes.)_

| Type                                                                                    | Applies? |
| --------------------------------------------------------------------------------------- | -------- |
| 🚨 _Breaking change_ (fix or feature that would cause existing functionality to change) |          |
| ✨ _New feature_ (non-breaking change that adds functionality)                          |        X  |
| 🐛 _Bug fix_ (non-breaking change that fixes an issue)                                  |          |
| ♻️ _Refactoring_ (internal change to codebase, without changing functionality)          |          |
| 🚦 _Test update_ (change that _only_ adds or modifies tests)                            |          |
| 📚 _Documentation update_ (change that _only_ updates documentation)                    |          |
| 📦 _Dependency update_ (change that updates a dependency)                               |          |
| 🔧 _Internal_ (change that _only_ affects developers or continuous integration)         |          |

## Checklist

_(Complete each of the following items for your pull request. Indicate that you have completed an item by changing the `[ ]` into a `[x]` in the raw text, or by clicking on the checkbox in the rendered description on GitHub.)_

Before opening your pull request:

- [x] I have performed a self-review of my changes.
  - Check that all changed files included in this pull request are intentional changes.
  - Check that all changes are relevant to the purpose of this pull request, as described above.
- [x] I have added tests for my changes, if applicable.
  - This is **required** for all bug fixes and new features.
- [x] I have updated the project documentation, if applicable.
  - This is **required** for new features.
- [x] I have updated the project Changelog (this is required for all changes).
- [x] If this is my first contribution, I have added myself to the list of contributors.

After opening your pull request:

- [x] I have verified that the pre-commit.ci checks have passed.
- [x] I have verified that the CI tests have passed.
- [x] I have reviewed the test coverage changes reported by Coveralls.
- [x] I have [requested a review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) from a project maintainer.

## Questions and Comments

_(Include any questions or comments you have regarding your changes.)_

david-yz-liu commented on python_ta/checkers/function_parameter_not_mentioned_checker.py:
```diff
@@ -0,0 +1,59 @@
+    msgs = {
+        "C9960": (
+            "The parameter '%s' is not mentioned in the docstring",
+            "function-parameters-not-mentioned",
```
Overall, please rename to `unmentioned-parameter`. This will have changes throughout.

(This fits the naming convention of most pylint errors, which seem to be of the form <adjective> <noun>. Note that "parameter" implies "function", so the latter word is unnecessary in the name.)

david-yz-liu commented on python_ta/checkers/function_parameter_not_mentioned_checker.py:
```diff
@@ -0,0 +1,59 @@
+
+class FunctionParameterNotMentionedChecker(BaseChecker):
+    """
+    A class checker to check if every function parameter is mentioned by name within it's the docstring.
```
"A class checker" -> "A class"

"within it's the docstring" -> "within the function's docstring"

david-yz-liu commented on python_ta/checkers/function_parameter_not_mentioned_checker.py:
```diff
@@ -0,0 +1,59 @@
+class FunctionParameterNotMentionedChecker(BaseChecker):
+    """
+    A class checker to check if every function parameter is mentioned by name within it's the docstring.
+    By default, this checker will be disabled.
```
"will be" -> "is"

david-yz-liu commented on python_ta/checkers/function_parameter_not_mentioned_checker.py:
```diff
@@ -0,0 +1,59 @@
+        )
+    }
+
+    def __init__(self, linter: Optional[PyLinter] = None) -> None:
```
Since this initializer just delegates to the parent class, you can omit its definition here

david-yz-liu commented on python_ta/checkers/function_parameter_not_mentioned_checker.py:
```diff
@@ -0,0 +1,59 @@
+    # Helper Function
+    def _check_parameters(self, docstring: str, parameters: list[str], node: nodes.NodeNG) -> None:
+        """Check if every parameter is mentioned in the docstring"""
+        words = {word.strip(".,:") for line in docstring.split("\n") for word in line.split()}
```
Overall it's a bit brittle to just remove a few specific punctuation marks, but I do like the instinct. A better approach is to use https://docs.python.org/3/library/string.html#string.punctuation

david-yz-liu commented on python_ta/checkers/function_parameter_not_mentioned_checker.py:
```diff
@@ -0,0 +1,59 @@
+        """Check if every parameter is mentioned in the docstring"""
+        words = {word.strip(".,:") for line in docstring.split("\n") for word in line.split()}
+        for parameter in parameters:
+            if not (words and parameter in words):
```
this condition is more complex than it needs to be; `parameter not in words` is fine

david-yz-liu commented on python_ta/checkers/function_parameter_not_mentioned_checker.py:
```diff
@@ -0,0 +1,59 @@
+    def visit_functiondef(self, node: nodes.FunctionDef) -> None:
+        """Visit a function definition"""
+        docstring = node.doc_node.value if node.doc_node and node.doc_node.value else ""
+        self._check_parameters(self._strip_docstring_of_doctest(docstring), node.argnames(), node)
```
You should provide a screenshot of what the error message looks like.

In particular, passing in `node` isn't ideal; this will cause the entire function definition to be highlighted, but really we should only highlight the specific parameter that isn't mentioned.

So instead of using `node.argnames()`, you'll need to figure out how to iterate over the function's parameters in the AST.

david-yz-liu commented on python_ta/checkers/function_parameter_not_mentioned_checker.py:
```diff
@@ -0,0 +1,59 @@
+        start_index = docstring.find(">>>")
+        contains_doctest = start_index != -1
+        if contains_doctest:
+            return docstring[:start_index]
```
Good to exclude doctests, but this is also a bit brittle, as it's possible to mix doctest examples and other text.

A better approach is to use `DoctestParser.parse`: https://docs.python.org/3/library/doctest.html#doctestparser-objects.

david-yz-liu reviewed (commented):
@CulmoneY nice work. In addition to the inline comments I left, please make sure to (1) add an example under the `examples` folder, and (2) update the documentation to describe this new checker. There's an existing checker for "missing docstring" or something similar, and you can put your new entry under that.

CulmoneY commented:
Hi David,
I wanted to inform you about an issue with the test `test_examples_files_pyta[examples/custom_checkers/c9960_unmentioned_parameter.py]`. The problem occurred because the checker for c9960 was disabled by default.

To resolve this, I have modified the test cases to explicitly enable the c9960 checker, ensuring that the test now runs as expected.

david-yz-liu commented on python_ta/checkers/function_parameter_not_mentioned_checker.py:
```diff
@@ -0,0 +1,56 @@
+
+    # Helper Function
+    def _check_parameter(self, docstring: str, parameter: str, node: nodes.NodeNG) -> None:
+        """Check's if parameter is mentioned in the docstring"""
```
"Check's if" -> "Return whether"

david-yz-liu commented on python_ta/checkers/function_parameter_not_mentioned_checker.py:
```diff
@@ -0,0 +1,56 @@
+from pylint.lint import PyLinter
+
+
+class FunctionParameterNotMentionedChecker(BaseChecker):
```
Rename this class, and the file itself, to align with the new name of the check. (Note that this also applies to the test files as well; generally speaking, a "renaming" change can actually affect a lot of places.)

david-yz-liu commented on tests/test_custom_checkers/test_function_parameter_not_mentioned_checker.py:
```diff
@@ -0,0 +1,302 @@
+    def setUp(self):
+        self.setup_method()
+
+    # TODO: The follwoing cases need to change, the node is no longer a function node, but instead a parameter node
```
It's not clear what this TODO is referring to.

david-yz-liu commented on tests/test_custom_checkers/test_function_parameter_not_mentioned_checker.py:
```diff
@@ -0,0 +1,302 @@
+        with self.assertNoMessages():
+            self.checker.visit_functiondef(function_node)
+
+    def test_multiple_missing_parameters(self) -> None:
```
This test name and description does not align with what's actually being tested.

david-yz-liu commented on tests/test_custom_checkers/test_function_parameter_not_mentioned_checker.py:
```diff
@@ -0,0 +1,302 @@
+        """Test the checker on a function with multiple missing parameters"""
+        src = '''
+                def f(x: int, y: int) -> int:
+                    """Both Parameters not mentioned
```
don't capitalize "Parameters"

CulmoneY commented:
Hey David,

I wanted to update you on the latest changes. I've ensured all functionalities we discussed in the meeting are working as expected. Specifically, all punctuation is now replaced by a space instead of being removed.

However, I noticed an issue where if a function parameter contained punctuation, the checker would incorrectly flag it. To fix this, I’ve updated the implementation to remove all punctuation except for the punctuation that is part of the parameter itself.

There is a rare edge case where the parameter name could be followed by punctuation that is part of the name, but I believe this scenario is extremely unlikely to occur.

david-yz-liu commented on python_ta/checkers/unmentioned_parameter_checker.py:
```diff
@@ -0,0 +1,61 @@
+
+    def _make_translation_table(self, parameter: str) -> dict:
+        """Return a translation table that removes all punctuation, expect for the punctuation within parameter"""
+        punctuation_in_parameter = {char for char in parameter if char in string.punctuation}
```
Oh and as a related note, you've also correctly identified that this means that extra underscores (e.g. at the end of a word) won't be stripped away when processing the docstring. I think that's okay.
